Jenkinsfile

```diff
@@ -0,0 +1,127 @@
+pipeline {
+    agent {
+        dockerfile {
+            filename 'build.Dockerfile'
+            args '-v /var/run/docker.sock:/var/run/docker.sock'
+        }
+    }
+
+    environment {
+        IMAGE_CREDS_JENKINS_ID = 'aks-datavault'
+        IMAGE_REGISTRY = 'datavault.azurecr.io'
+        IMAGE_REPOSITORY_WEBAPP = 'webapp'
+        IMAGE_REPOSITORY_BROKER = 'broker'
+        IMAGE_REPOSITORY_WORKER = 'worker'
+        DEPLOYMENT_JOB = '../datavault-infra/master'
+        DEPLOYMENT_ENV = 'dev'
+    }
+
+    options {
+        ansiColor('xterm')
+        timestamps()
+    }
+
+    stages {
+        stage('general linting') {
+            steps {
+                sh 'pre-commit install'
+                sh 'pre-commit run --all-files --verbose || true'
+            }
+        }
+
+        stage('dockerfile linting') {
+            agent {
+                docker {
+                    image 'hadolint/hadolint:latest-debian'
+                }
+            }
+            steps {
+                sh 'hadolint *.Dockerfile | tee -a hadolint_lint.txt'
+            }
+            post {
+                always {
+                    archiveArtifacts 'hadolint_lint.txt'
+                }
+            }
+        }
+
+        stage('test') {
+            steps {
+                sh 'mvn clean test -U --batch-mode'
+            }
+        }
+
+        stage('build images') {
+            steps {
+                sh 'docker build -t datavault/maven-build:latest -f maven.Dockerfile .'
+                sh 'docker build -t $IMAGE_REPOSITORY_WEBAPP:latest -f webapp.Dockerfile .'
+                sh 'docker build -t $IMAGE_REPOSITORY_BROKER:latest -f broker.Dockerfile .'
+                sh 'docker build -t $IMAGE_REPOSITORY_WORKER:latest -f worker.Dockerfile .'
+            }
+        }
+
+        stage('push images') {
+            when {
+                branch "master"
+            }
+
+            steps {
+                withCredentials([usernamePassword(credentialsId: "$IMAGE_CREDS_JENKINS_ID", usernameVariable: 'IMAGE_REGISTRY_USERNAME', passwordVariable: 'IMAGE_REGISTRY_PASSWORD')]) {
+                    sh 'docker login $IMAGE_REGISTRY --username $IMAGE_REGISTRY_USERNAME --password $IMAGE_REGISTRY_PASSWORD'
+                }
+
+                script {
+                    def properties = readProperties(file: 'version.properties')
+                    version = "${properties.version}-${currentBuild.startTimeInMillis}.${currentBuild.number}"
+
+                    def webappImages = [
```

Comment:
It feels like there's lots of repetition here where the only difference between these blocks is which image repository we're using. It might be worth seeing if these can be pulled out into a function that can be reused

Reply:
This sounds like a micro optimisation that would cost more to refactor and implement than would be gained by doing. I agree it's a bit ugly right now, but this isn't never going to get any bigger since the number of images is static for now and the foreseeable future.